### ISSUE TITLE
Request: Tasker support

### ISSUE THREAD

**Issue by Cris--** [mkulesh/onpc#184]
First: Great app! A way better navigation then the original Onkyo Controller app!

Request: Tasker support to use the Receiver for morning alarm for the whole family.. 😋🤡😈

Or automatically switch inputs on certain situations.

..and many more.

**Comment by mkulesh** [comment]
Hi @Cris--, I have no experience with Tasker, but the app is already able to process some commands from Tasker given in the intent. The technical implementation of such a processing is already done within the issue https://github.com/mkulesh/onpc/issues/155.
@qpkorr develops currently Tasker scripts (for example, for voice control of the app).

@qpkorr, could you please share your experience with Tasker for @Cris--? I think we shall start to commit the documentation about indent as discussed via emails some days ago. 

@Cris--, could you please write me on mikhail.kulesh@gmail.com and i will provide you more information about this feature.

**Comment by qpkorr** [comment]
Hi @Cris-- (and @mkulesh) - yep, you weren't the only one to think those two thoughts @Cris-- - a great app - how about some love for Tasker :)  @mkulesh has been coaching me to help implement message scripts on the onpc side, whilst I've made a few tasker scripts to send intents to onpc.  The intents send ISCP commands - the Onkyo protocol that @mkulesh has been implementing - so it's somewhat low level at this stage - but there's plans to make it easier, especially now that @mkulesh has added Shortcuts, which enable storing sequences of commands which use the same message scripts.  In short, the sky is the limit really - and the functionality you need is already there. 

I've got some tasker profiles and tasks I can share as a start - they're far from perfect but they work.  In particular, since I first wrote them, Joao has added support for passing more than two parameters when calling tasks - which I should most probably be refactoring for.  We'd already planned to document the tasker side of things soon as my voice control scripts became stable just last week - so your request came at a very good time!  Stay tuned!

**Comment by Cris--** [comment]
Is Intent the same as Activities? No or...?
Possible via dynamic Activities maybe?
I'm an programming noob but in Tasker not ^^

**Comment by qpkorr** [comment]
What are intents?  From here: https://tasker.joaoapps.com/userguide/en/intents.html
"Intents are Android's main method for allowing apps to communicate with each other and share data. Intents are for advanced users."

Whilst intents are a little complex, hopefully that complexity can be hidden sufficiently well behind tasker tasks I've already written.  And more significantly, whilst the app currently supports only the ISCP messages via intents, @mkulesh and I have been discussing adding support for running Shortcuts from Tasker - which should make things a lot simpler.

As for dynamic Activities - I'm not really sure what you mean. By activities do you mean tasks? If so yes, I guess we use a dynamic task to send an intent to onpc. They're a little complex, but ok once you get the hang of them I expect :)

**Comment by Cris--** [comment]
Please share a screenshot from Tasker/send intent or update manual with a example like for e.g. Power On.

![Screenshot_20201010-120834951](https://user-images.githubusercontent.com/7994691/95652509-d01eaf00-0af1-11eb-97fb-05ee521c1d10.jpg)

**Comment by Cris--** [comment]
Can't get it work with onpc 1.17

![Screenshot_20201010-165341751](https://user-images.githubusercontent.com/7994691/95658286-8cd83680-0b19-11eb-9672-b44052c31491.jpg)

**Comment by qpkorr** [comment]
I'm not sure what Ziel means, but I think it should be Activity. Will provide a screenshot later - having technical problems with them at present

**Comment by qpkorr** [comment]
This works for me :)
![Screenshot_20201011-022612_Tasker](https://user-images.githubusercontent.com/347545/95670102-49420300-0bd3-11eb-8c31-01e0c0bac536.jpg)

**Comment by mkulesh** [comment]
Hi @Cris--,
I updated https://github.com/mkulesh/onpc/blob/master/doc/Tasker/README.md based on the input from @qpkorr 
Please try to configure your intent as shown in the screenshot above.
@qpkorr I also moved the "Voice Control" section into the README.md. Could you please review whole README.md file?

**Comment by Cris--** [comment]
> I'm not sure what Ziel means, but I think it should be Activity. Will provide a screenshot later - having technical problems with them at present

"Ziel" means _Source_ and the source _Activity_ make no sense with onpc (opens onpc app then). This must be go in the background.

Great now with this Tasker/onpc possibilities but now we need an easier solution for humans, not for devs only!  ;)
What i understand now is i can send xml scripts via variable or send directly like _%iscpCmd PWR;OFF_
For me the last is relevant. Simple commands to trigger on/off the AV Receiver, switch the source, or set a specific volume, etc.
But can't get it work with the above command.
The Tasker project OnpcVoiceControl.prj.xml is nice to see what can be done but for me completely overdose. 

I think i need a little more help. ^^

**Comment by qpkorr** [comment]
> "Ziel" means Source and the source Activity make no sense with onpc (opens onpc app then). This must be go in the background.

Fair enough - thus far, I've only used it open the onpc app in the foreground, mainly in order to search for music which I then play manually.  You're quite right that a background approach would be good for other use cases (eg volume control that I'm currently implementing).  I've no idea what's possible in this regard, or how much work it would be in onpc.

**Comment by mkulesh** [comment]
Hi @Cris-- 
if you put the script like this:
<?xml version="1.0" encoding="utf-8"?>
<onpcScript host="10.49.0.7" port="60128" zone="0" tab="MEDIA">
  <send cmd="PWR" par="QSTN" wait="PWR"/>
  <send cmd="PWR" par="01" wait="PWR" resp="01"/>
</onpcScript>
in field "Daten", does it work for you or not?

If yes, you can reduce it as follows:
<?xml version="1.0" encoding="utf-8"?>
<onpcScript host="" port="" zone="0" tab="MEDIA">
  <send cmd="PWR" par="01" wait="PWR"/>
</onpcScript>

**Comment by Cris--** [comment]
PWR 01 did work, PWR 00 not. Tried also with _wait_ and 500 or 1000. Onkyo TX-NR575E

**Comment by mkulesh** [comment]
Hi @Cris-- 
I installed the Tasker and configured two tasks "Power On" and "Power Off", see new revision of https://github.com/mkulesh/onpc/blob/master/doc/Tasker/README.md.
Both scripts work fine for me. Please configure it as shown in the screenshots, but keep host and port empty as in your example above. 

In order to work stable, a sequence shall contain _at least three commands_:
- wait until connection is established
```
<send cmd="NA" par="NA" wait="NRI"/>
```
- request the actual power state
```
<send cmd="PWR" par="QSTN" wait="PWR"/>
```
- change the power state if necessary

**Comment by qpkorr** [comment]
Ah, that explains why the README had changed between when I refreshed it to review it, and when I looked on github again :)
Interesting that you're using the Tasker Action App/Launch App, where I've been using System/Send Intent.  If intending to open the app, Launch App looks simpler thus better. If running the command in the background is possible, I guess Send Intent might be required.  If Launch App is going to be the recommended method, the description of Intents can probably be removed.

**Comment by mkulesh** [comment]
Hi @qpkorr 
please add this useful comment as well as your previous comments  to the README!

**Comment by Cris--** [comment]
@mkulesh Power Off works now!

**Comment by Cris--** [comment]
Ok, nice. Here my **Power On + Tuner (FM)** input script:

```
<?xml version="1.0" encoding="utf-8"?>
 <onpcScript host="" port="" zone="0" tab="LISTEN">
  <send cmd="PWR" par="01" wait="PWR"/>
  <send cmd="SLI" par="24" wait="SLI"/>
</onpcScript>
```

I think we are now on the right way! Big **THANKS!** ^^

**Comment by mkulesh** [comment]
Hi @Cris-- 
I propose to add state requests before PWR and SLI commands in order to make script more stable. The problem is that when the tasker starts the app with the script, the app performs a lot of background communication with the receiver in order to obtain complete state. If you sends the commands directly it may conflict with things that app does in background. My suggested script is:
```
<?xml version="1.0" encoding="utf-8"?>
 <onpcScript host="" port="" zone="0" tab="LISTEN">
  <send cmd="NA" par="NA" wait="NRI"/>
  <send cmd="PWR" par="QSTN" wait="PWR"/>
  <send cmd="PWR" par="01" wait="PWR"/>
  <send cmd="SLI" par="QSTN" wait="SLI"/>
  <send cmd="SLI" par="24" wait="SLI" resp="24"/>
  <send cmd="PRS" par="QSTN" wait="PRS"/>
  <send cmd="PRS" par="01" wait="PRS"/>
</onpcScript>
```

**Comment by Cris--** [comment]
Maybe, maybe not. After my tests now ( min 30x play, switch, etc ) no problems at the moment. All fine!

**Comment by mkulesh** [comment]
Hi @Cris-- 
ok, no problem. In the case if you will observe any problems with your script please try my suggestion in order to resolve it.

**Comment by Cris--** [comment]
for the **Data** field in T**asker > Send** Intent can you support also _file://_ & _content://_ with the onpc app please?
So we can point directly to xml files. Makes a lot easier!

**Comment by mkulesh** [comment]
Hi @Cris-- 
I do not plan to add such a functionality since it requires additional permissions in order to read file storage. One of the advantage of my app is a really small list of necessary permissions and I like to keep it. 
In order to make the script creation easier, I implemented https://github.com/mkulesh/onpc/issues/190.
If you have some shortcuts in the app, you can just copy the script related to the shortcut and paste it in Tasker.

**Comment by mkulesh** [comment]
Hi @Cris-- 
"Copy to clipboard" for shortcuts is now available in debug build. If you like to try it, please install
https://github.com/mkulesh/onpc/raw/autobuild/autobuild/onpc-v1.18-debug.apk

**Comment by mkulesh** [comment]
Released in version 1.18
